ArduSub/control_althold.cpp

```diff
@@ -95,21 +95,37 @@ void Sub::althold_run()
         }
     }
 
+    // Hold actual position until zero derivative is detected
+    static bool holdIt = true;
+    // Get last user velocity direction to check for zero derivative points
+    static float lastVel = 0;
     if (fabsf(channel_throttle->norm_input()-0.5f) > 0.05f) { // Throttle input above 5%
         // output pilot's throttle
         attitude_control.set_throttle_out(channel_throttle->norm_input(), false, g.throttle_filt);
         // reset z targets to current values
-        pos_control.relax_alt_hold_controllers(motors.get_throttle_hover());
+        pos_control.relax_alt_hold_controllers(motors.get_throttle_hover(), false);
+        holdIt = true;
+        lastVel = inertial_nav.get_velocity_z();
     } else { // hold z
 
         if (ap.at_bottom) {
-            pos_control.relax_alt_hold_controllers(motors.get_throttle_hover()); // clear velocity and position targets, and integrator
+            pos_control.relax_alt_hold_controllers(motors.get_throttle_hover(), false); // clear velocity and position targets, and integrator
             pos_control.set_alt_target(inertial_nav.get_altitude() + 10.0f); // set target to 10 cm above bottom
         } else if (rangefinder_alt_ok()) {
             // if rangefinder is ok, use surface tracking
             float target_climb_rate = get_surface_tracking_climb_rate(0, pos_control.get_alt_target(), G_Dt);
             pos_control.set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
         }
+
+        // Detects a zero derivative
+        // When detected, move the altitude set point to the actual position
+        // This will avoid any problem related to joystick delays
+        // or smaller input signals
+        if(holdIt && lastVel*inertial_nav.get_velocity_z() < 0 ) {
```

Comment:
Does float multiplication cost significantly more cycles than tracking sign?

Reply:
not sure, a benchmark is necessary.

Reply:
I think it must be so, and we should track/compare only the sign (boolean).

Reply:
Done.